crates/context/primitives/src/client.rs

```diff
@@ -46,37 +59,232 @@ impl ContextClient {
+                    yield k.public_key();
+                }
+            }
+        }
     }
 
     pub async fn has_member(
         &self,
         context_id: &ContextId,
         public_key: &PublicKey,
-        is_present: bool,
     ) -> eyre::Result<bool> {
-        todo!()
+        let handle = self.datastore.handle();
+
+        let key = ContextIdentityKey::new(*context_id, *public_key);
+
+        Ok(handle.has(&key)?)
     }
 
-    pub async fn update_application_id(
+    pub async fn update_application(
         &self,
         context_id: &ContextId,
         application_id: &ApplicationId,
         identity: &PublicKey,
     ) -> eyre::Result<()> {
-        todo!()
+        let (sender, receiver) = oneshot::channel();
+
+        self.context_manager
+            .send(ContextMessage::UpdateApplication {
+                request: UpdateApplicationRequest {
+                    context_id: *context_id,
+                    application_id: *application_id,
+                    public_key: *identity,
+                },
+                outcome: sender,
+            })
+            .await
+            .expect("Context manager mailbox not to be dropped");
+
+        receiver
+            .await
+            .expect("Context manager not to drop response channel")
+    }
+
+    pub async fn invite_member(
+        &self,
+        context_id: &ContextId,
+        inviter_id: &PublicKey,
+        invitee_id: &PublicKey,
+    ) -> eyre::Result<Option<ContextInvitationPayload>> {
+        let handle = self.datastore.handle();
+
+        let Some(context_config) = handle.get(&ContextConfigKey::new(*context_id))? else {
+            return Ok(None);
+        };
+
+        let Some(ContextIdentityValue {
+            private_key: Some(requester_secret),
+            ..
+        }) = handle.get(&ContextIdentityKey::new(*context_id, *inviter_id))?
+        else {
+            return Ok(None);
+        };
+
+        let nonce = self
+            .external_client
+            .query::<ContextConfigEnv>(
+                context_config.protocol.as_ref().into(),
+                context_config.network.as_ref().into(),
+                context_config.contract.as_ref().into(),
+            )
+            .fetch_nonce(
+                context_id.rt().expect("infallible conversion"),
+                inviter_id.rt().expect("infallible conversion"),
+            )
+            .await?
+            .ok_or_eyre("The inviter doesen't exist")?;
+
+        self.external_client
+            .mutate::<ContextConfigEnv>(
+                context_config.protocol.as_ref().into(),
+                context_config.network.as_ref().into(),
+                context_config.contract.as_ref().into(),
+            )
+            .add_members(
+                context_id.rt().expect("infallible conversion"),
+                &[invitee_id.rt().expect("infallible conversion")],
+            )
+            .send(requester_secret, nonce)
+            .await?;
```

Comment:
with the new external client abstractions, all of this.. should just become:

```rust
let external_config = self.context_config(context_id);

let external_client = self.external_client(context_id, &external_config);

external_client.config().add_members().await?;
```

then you can use `external_config` for the invitation payload